yktrace/src/hwt/mapper.rs

```diff
@@ -108,6 +108,61 @@ impl HWTMapper {
         }
     }
 
+    /// Collapse consecutive control point blocks (including any unmappable internals) into one
+    /// control point entry. Unmappable holes exist because the generated control point calls out to
+    /// Rust code for which we have no IR.
+    ///
+    /// For example, if the trace is:
+    ///
+    /// ```ignore
+    /// [bb1, bb2, bb3, ctrlp, ctrlp, ?, ctrlp, bb4]
+    /// ```
+    ///
+    /// (where `bbN` is a mappable block, `ctrlp` is a block in the control point, and `?` is
+    /// unmappable blocks), then this function would canonicalise the trace to:
+    ///
+    /// ```ignore
+    /// [bb1, bb2, bb3, ctrlp, bb4]
+    /// ```
+    fn canonicalise_control_point(irblocks: &mut Vec<IRBlock>) {
+        let ctrlp_sym_c = CString::new(CONTROL_POINT_SYM).unwrap();
+        assert!(irblocks[0].func_name == ctrlp_sym_c);
+        let mut in_ctrlp = true;
+        let mut ctrlp_start_idx = Some(0);
+        let mut delete_ranges = Vec::new();
+        for (ib_idx, ib) in &mut irblocks.iter().enumerate() {
+            if in_ctrlp {
+                // If we encounter an unmappable hole, or another block from CONTROL_POINT_SYM,
+                // then we remain inside the control point. Otherwise we have just exited the
+                // control point.
+                if !(ib.is_unmappable() || ib.func_name == ctrlp_sym_c) {
```

Comment:
Might this be clearer as:

```rust
if !ib.is_unmappable() || ib.func_name != ctrlp_sym_c { ... }
```

Reply:
Yes, but you meant `&&` I think?

[One of the few things I remember from college education: [De Morgan's Laws](https://en.wikipedia.org/wiki/De_Morgan%27s_laws) :laughing:]

Reply:
Fixed in f9c355c.